src/mod/server.mod/sasl.c

```diff
@@ -0,0 +1,716 @@
+  char error_msg[128]; /* snprintf() truncation should be tolerable */
+  int salt_plain_len, iter, j, ret;
+  char salt_plain[64]; /* atheme: Valid values are 8 to 64 (inclusive) */
+  unsigned int stored_key_len;
+  unsigned char stored_key[EVP_MAX_MD_SIZE];
+  char client_final_message_without_proof[1024];
+  unsigned char client_signature[EVP_MAX_MD_SIZE];
+  unsigned char client_proof[EVP_MAX_MD_SIZE];
+  char client_proof_b64[1024];
+  struct rusage ru1, ru2;
+
+  strlcpy(server_first_message, server_msg_plain, sizeof server_first_message);
+  for (word = strtok_r(server_msg_plain,  ",", &brkb);
+       word;
+       word = strtok_r(NULL, ",", &brkb)) {
+    switch (*word) {
+      case 'r':
+        if (
+#if OPENSSL_VERSION_NUMBER >= 0x1010008fL /* 1.1.0h */
+            CRYPTO_memcmp
+#else
+            memcmp
+#endif
+            (word + 2, nonce, (sizeof nonce) - 1)) {
+          sasl_error("AUTHENTICATE: server nonce != client nonce");
+          return -1;
+        }
+        server_nonce = word + 2;
+        break;
+      case 's':
+        salt_b64 = word + 2;
+        break;
+      case 'i':
+        i = word + 2;
+        break;
+      case 'e':
+        snprintf(error_msg, sizeof error_msg, "AUTHENTICATE: server error: %s", word + 2);
+	sasl_error(error_msg);
+        return -1;
+      default:
+        putlog(LOG_SERV, "*", "SASL: AUTHENTICATE warning: SCRAM Attribute ignored: %s", word);
+    }
+  }
+  if (!server_nonce) {
+    sasl_error("AUTHENTICATE: server nonce missing from SCRAM challenge");
+    return -1;
+  }
+  if (!salt_b64) {
+    sasl_error("AUTHENTICATE: salt missing from SCRAM challenge");
+    return -1;
+  }
+  if (!i) {
+    sasl_error("AUTHENTICATE: iteration count missing from SCRAM challenge");
+    return -1;
+  }
+  /* TODO: normalize(password)
+   * Eggdrop doesnt have support for utf8 normalization yet
+   * tcl also doesnt have it in core yet, only in tcllib
+   * We could use glib or something, but we dont want dependency bloat just
+   * for one function
+   */
+
+  /* ClientKey       := HMAC(SaltedPassword, "Client Key") */
+
+  use_cache = (!strcmp(sasl_password, last_sasl_password)) && (!strcmp(salt_b64, last_salt_b64)) && (!strcmp(i, last_i));
+
+  if (!use_cache) {
+    if ((salt_plain_len = b64_pton(salt_b64, (unsigned char*) salt_plain, sizeof salt_plain)) == -1) {
+      sasl_error("AUTHENTICATE: could not base64 decode salt");
+      return -1;
+    }
+    errno = 0;
+    iter = strtol(i, NULL, 10);
+    if (errno) {
+      snprintf(error_msg, sizeof error_msg, "AUTHENTICATE: strtol(%s): %s", i, strerror(errno));
+      sasl_error(error_msg);
+      return -1;
+    }
+
+    if (sasl_mechanism == SASL_MECHANISM_SCRAM_SHA_256)
+      digest = EVP_sha256();
+    else
+      digest = EVP_sha512();
+    digest_len = EVP_MD_size(digest);
+
+    ret = getrusage(RUSAGE_SELF, &ru1);
+    if (!PKCS5_PBKDF2_HMAC(sasl_password, strlen(sasl_password),
```

Comment:
What is the relation between PBKDF2 and SASL here? If PBKDF2 isn't loaded, does this become an issue?

Reply:
`PKCS5_PBKDF2_HMAC()` is an openssl 1.0.0+ function. sasl scram doesnt depend on pbkdf2.mod.